rts/motoko-rts/src/gc/generational/sanity_checks.rs

```diff
@@ -0,0 +1,183 @@
+//! Extensive sanity checks for geberational GC features.
+//! * Write barrier coverage by memory snapshot comparisons.
+//! * Memory sanity check, including a full heap scan.
+#![allow(dead_code)]
+
+use core::ptr::null_mut;
+
+use super::write_barrier::REMEMBERED_SET;
+use super::{Heap, Limits, Roots};
+use crate::mem_utils::memcpy_bytes;
+use crate::memory::{alloc_blob, Memory};
+use crate::types::*;
+use crate::visitor::{pointer_to_dynamic_heap, visit_pointer_fields};
+
+static mut SNAPSHOT: *mut Blob = null_mut();
+
+/// Take a memory snapshot. To be initiated after GC run.
+pub unsafe fn take_snapshot<M: Memory>(heap: &mut Heap<M>) {
```

Comment:
Not sure this is worth it, but I guess you could consider snapshoting to stable memory which might work with larger heaps without exhausting main memory. The verification would be much slower though, unless you do some caching of stable memory pages during verification.

Reply:
Maybe not worth the complexity...

Reply:
This is an interesting idea. As it is only for sanity checks, I may try this when there is need for for memory. The snapshot check is anyway very expensive in terms of runtime, and therefore probably only used for very specific tests in the longer term. Probably it will be worse with stable memory.